lib/src/Uncertainty/Algorithm/WeightedExperiments/TensorProductExperiment.cxx

```diff
@@ -0,0 +1,141 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/TensorProductExperiment.hxx"
+#include "openturns/Tuples.hxx"
+#include "openturns/Exception.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(TensorProductExperiment)
+
+static const Factory<TensorProductExperiment> Factory_TensorProductExperiment;
+
+typedef Collection< Point > PointCollection;
+
+/* Default constructor */
+TensorProductExperiment::TensorProductExperiment()
+  : WeightedExperimentImplementation()
+  , collection_(0)
+  , nodes_(0, 0)
+{
+  // Nothing to do
+}
+
+/* Constructor with parameters */
+TensorProductExperiment::TensorProductExperiment(const ExperimentPersistentCollection & collection)
+  : WeightedExperimentImplementation()
+  , collection_(0)
+  , nodes_(0, 0)
+{
+
+}
+
+/* Virtual constructor */
+TensorProductExperiment * TensorProductExperiment::clone() const
+{
+  return new TensorProductExperiment(*this);
+}
+
+/* String converter */
+String TensorProductExperiment::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << GetClassName()
+    << " name=" << getName()
+    << " collection=" << collection_;
+  return oss;
+}
+
+Bool TensorProductExperiment::hasUniformWeights() const
+{
+  return false;
+}
+
+/* Sample generation */
+Sample TensorProductExperiment::generateWithWeights(Point & weights) const
+{
+  if (!isAlreadyComputedNodesAndWeights_) computeNodesAndWeights();
+  weights = weights_;
+  return nodes_;
+}
+
+/* Compute the tensor product nodes and weights */
+void TensorProductExperiment::computeNodesAndWeights() const
+{
+  const UnsignedInteger dimension = collection_.getSize();
+  if (!(dimension > 0))
+    throw InvalidArgumentException(HERE) << "Error: expected a positive dimension, here dimension is " << dimension;
+  // Compute marginal nodes and weights
+  Collection<Sample> marginalNodesCollection;
+  Collection<Point> marginalWeightsCollection(dimension);
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    Point weights;
+    const Sample nodes(collection_[i].generateWithWeights(weights));
+    marginalNodesCollection.add(nodes);
+    marginalWeightsCollection.add(weights);
+  }
+  // Compute number of nodes for each dimension
+  // which depends on the marginal levels.
+  Indices marginal_size;
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    marginal_size[i] = marginalWeightsCollection[i].getDimension();
+  }
+  // Create index set
+  Tuples tuples(marginal_size);
+  IndicesCollection indices_collection(tuples.generate());
+  // Compute nodes and weights
+  const UnsignedInteger size = indices_collection.getSize();
+  Point weights(size);
+  Sample nodes(size, dimension);
```

Comment:
Here you make the assumption that all the marginal weighted experiments are of dimension 1, which is not mandatory for the algorithm to work

Reply:
Of course, but this is more tricky. I have to work a bit more on this topic.

Reply:
It should be there at the end of the day.